Reduce concurrency

cristiklein opened the pull request:
**What this PR does / why we need it**:

This PR serializes the part of apps which "changes" the Kubernetes API, e.g., by installing CRDs or webhooks. This should eliminate annoying "webhook timeout" errors that kept bothering us.

**Which issue this PR fixes** *(use the format `fixes #<issue number>(, fixes #<issue_number>, ...)` to automatically close the issue when PR gets merged)*: fixes #402 #391 #369 

**Public facing documentation PR** *(if applicable)*
<!-- https://github.com/elastisys/compliantkubernetes/pull/ -->

N/A

**Special notes for reviewer**:

Template validation before `helmfile apply` had to be removed. This is an unfortunate compromise with properly handling CRDs. Note, however, that Helmfile does template validation on install.

Note that, this PR build on #444 , hence, it is marked as Draft until the dependent PR is merged.

**Checklist:**

- [ ] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [X] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [X] is completely transparent, will not impact the workload in any way.
    - [ ] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.

**Pipeline config** *(if applicable)*
If you change some config options (e.g. add/rename variable or change the default value) you may need to update the config used by the pipeline in `pipeline/config`.

<!--
Here are the commit prefixes and comments on when to use them:
all: (things that touch on more than one of the areas below, or don't fit any of them)
apps: (changes to the applications running in both/all clusters)
apps sc: (changes to applications in the service cluster)
apps wc: (changes to applications in the workload cluster)
docs: (documentation)
tests: (test related changes)
pipeline: (the pipeline)
config: (configuration, e.g. add/remove/rename a parameter, this is not for changes to the default values for an application that would go into `apps [sc/wc]`)
bin: (changes to binaries or scripts used manage ck8s)
release: (anything release related)

Example commit prefix usage:

git commit -m "docs: Add instructions for how to do x"
-->

simonklb commented on helmfile/02-gatekeeper.yaml:
```diff
@@ -27,6 +27,7 @@ releases:
   labels:
     app: gatekeeper
   chart: upstream/gatekeeper
+  disableValidationOnInstall: true
```
Maybe worth adding a comment with the rational here? It would be nice to keep track of if this was fixed upstream at some point so that you could still keep validation with CRD-installing charts.

simonklb commented on helmfile/01-prometheus-operator.yaml:
```diff
@@ -27,8 +27,7 @@ releases:
   missingFileHandler: Error
-  needs:
-  - cert-manager/cert-manager
+  disableValidationOnInstall: true
```
Maybe worth adding a comment with the rational here? It would be nice to keep track of if this was fixed upstream at some point so that you could still keep validation with CRD-installing charts.

simonklb commented on helmfile/01-prometheus-operator.yaml:
```diff
@@ -27,8 +27,7 @@ releases:
   version: 12.8.0
   missingFileHandler: Error
-  needs:
-  - cert-manager/cert-manager
```
nit: This feels like it should live in one of the "lift up" commits

simonklb commented on helmfile/04-cert-manager.yaml:
```diff
@@ -27,6 +27,7 @@ releases:
   chart: upstream/cert-manager
   version: v1.1.0
   missingFileHandler: Error
+  disableValidationOnInstall: true
```
The commit message seems to be off on this one? I guess this should live in this commit: 8a7fe136faf74bb4251716cb1e6d5ef3717d5662

simonklb commented on WIP-CHANGELOG.md:
```diff
@@ -0,0 +1,3 @@
+Changed:
```
We list the requirements in the README as well. Not sure if this is the best of ideas, but for now update it there as well.

simonklb reviewed (changes requested):
I'm OK with this change. The extra time added is probably slim to none compared to the complexity it removes.

Mostly nits.

cristiklein commented:
@simonklb I think I addressed all issues you raised and pinky promise to be more careful about commit granularity in the future. (Although, TBH, I kind of imagine that this will all be squashed.) PTAL

simonklb commented:
> @simonklb I think I addressed all issues you raised and pinky promise to be more careful about commit granularity in the future. (Although, TBH, I kind of imagine that this will all be squashed.) PTAL

Thanks! :D Squashing it is fine, I like that you commited it in patches though, made reviewing easy, except for the parts that was included in the wrong context haha.

cristiklein commented:
@simonklb Can I get your seal of approval? :smile: :whale: ?

simonklb commented:
> @simonklb Can I get your seal of approval? smile whale ?

Could you re-point this from main to the other branch so that I can just see the changes done in this branch?

cristiklein commented:
@simonklb Sorry, wrong PR. :smile: 

Can I get your seal of approval on #444 ? After that, I will rebase this branch and ask for seal of approval on this PR too.

simonklb commented:
Please request a re-review once it's done!

simonklb reviewed (approved):
Nice!

I like that the blackbox dependency is vendored here now too. I'm curious how we should track that everything is being vendored though so that new home-made charts are not introduced without vendoring. Is it possible to disable runtime dependency fetching and building in helmfile perhaps?

cristiklein commented:
> I like that the blackbox dependency is vendored here now too. I'm curious how we should track that everything is being vendored though so that new home-made charts are not introduced without vendoring. Is it possible to disable runtime dependency fetching and building in helmfile perhaps?

We could add `--skip-deps` in `apply-(sc|wc).sh` if it bites us.

pettersv reviewed (approved):
LGTM!
